### DISCUSSION TITLE
Dynamic map bias

### DISCUSSION THREAD

**Discussion by u1ps** [etlegacy/etlegacy#2067] [Ideas]
Using total amount of maps played and wins in map bias calculation will eventually end up in static number once total number goes big enough. 

I think using last played maps for the data is more accurate as it will display recent state of map in the server. It will update if one of the teams would start performing better for whatever reason.

For example gathering info only from past 100 maps gives around 0,5% fluctuation after each map and make the number more dynamic and reflect current state of map better.

For example 1000 maps fives fluctuation about 0,05%. Once the number goes big enough it becomes very static and will not show current state of map but rather static number gathered over long period of time.

This graphic below will illustrate how the number will become linear and harder to change

![qqqqqqq](https://user-images.githubusercontent.com/99591943/180036317-b98934d0-da53-4546-88a6-7d1ee183bd04.png)

**Comment by u1ps** [comment]
Total amount of wins will eventually end up in very static line. Downscaling amount of wins to certain value will make map bias more dynamic. Simulate every game as it was x-th game (100th, 200th, 500th etc). Using value of 100 will cause every game around 0,5% change in map bias, 500 will cause change of 0,1%. And it start and continue in loop not losing dynamic effect. Also eliminating big fluctuation in start of data and retaining fluctuation in huge numbers.

Example:

Current balance is allies: 250 wins , axis 150 wins. Match limit: 400
After match           allies; 251wins , axis 150 wins. total value = 401 (over limit)

downscale to 400

Allies wins: 250,3741, Axis wins: 149,6259 



**1. Downscale (Orange) vs no downscale (Blue) (default)**

![10090](https://user-images.githubusercontent.com/99591943/180614993-b9274831-fd3d-4b9a-8234-262d5149a2e7.png)

Blue line consists of 5000 events, each event simulating 1 match with 50% win rate. Moving towards static number.

Orange line is same data but after 100th match each next match is dowscaled to 100. So each match from that point on behaves as it is 101th match. 

Those fluctuations are actually one team winning more but due to huge amount of data not making any change in bias (blue vs orange).

**2.1. Downscale to 100**

![100pp](https://user-images.githubusercontent.com/99591943/180651097-d8365ca6-4aee-430e-a901-099845358c46.png)

Blue is default, starting from 0 wins to either team. In start big fluctuations up to 5% and rapidly going down and eventually end up in no fluctuation. (No more change in bias)
Orange is downscale to 100, starting from 50wins to allies and 50wins to axis and then downscaled each game. Fluctuation value always around 0,5%. (Always change in bias)

**2.2. Zoom in on 100th event**

![100p](https://user-images.githubusercontent.com/99591943/180651054-5f8cfdac-93f8-462d-bc1c-7595d32b2a80.png)

Matches with games around 100 as can be seen on graph. Fluctuation value always around 0,5%. And from this point on same characteristics keep repeating.

**3.1. Downscale to 500**

![500pp](https://user-images.githubusercontent.com/99591943/180651137-40f8cd61-f68b-48d4-acec-b6a79be361f5.png)

Orange is downscale to 500, starting from 250wins to allies and 250wins to axis and then downscaled each game. Fluctuation value always around 0,1%. (Always change in bias, but smaller compared to downscale 100).

**3.2. Zoom in on 500th event**

![500p](https://user-images.githubusercontent.com/99591943/180651141-2662aced-a9be-4cbf-a7f7-5b17c0757bf9.png)

Matches with games around 500 as can be seen on graph. fluctuation value always around 0,1%. And from this point on same characteristics keep repeating.

**4.1. Change in win rate, downscale to 500**

![500change](https://user-images.githubusercontent.com/99591943/180618182-43097dd8-4104-400a-b2e8-002aa726aa6d.png)

Here is a data of 5000 events 

- 1- 2500 win chance of 60% for one team, 

- 2501 - 5000 swapped around , 60% win chance to other team. 

Orange line is downscale to 500. It can change and reflect true events much more accurately. Compared to huge chunk of data that cant even make it to positive ratio last 2500 games have been +10% chance to win. 

**4.2.1. Change in win rate, downscale to 100**

![100change](https://user-images.githubusercontent.com/99591943/180618177-a575c364-3b61-4c60-b65a-254be0c26083.png)

Downscale to 100, fluctuation is bigger but also adjustment happens much more quickly which can be seen in first matches (1-100). At 250, where win ratio swap happen it can reflect new accurate value in about 100 matches (which is perhaps around month in time period). it does hit +10% quite quickly and keeps fluctuating there as pretty versatile to change.

Also fluctuations from start (simulating perhaps new map) are also not that big, rather gradually moving towards real value from balanced state compared to old blue data line which has very big ranges in the start as it starts to fixing around value and increase precision on every continues match

Old data chunk cant even make it to positive ratio yet last 2500 games have been +10% chance to win. 

fixed on 100 managed to swap around in about 100 games while reflecting the 20% change in input yet old data line only raised few %


**4.2.2. Change in win rate, downscale to 100, zoom in on 2500th event**

![2500swap](https://user-images.githubusercontent.com/99591943/180618815-833227fa-f362-4f7e-b679-2b0157fdaa75.png)

Such downscaling eliminates huge bias changes in new maps as there is not enough win/lose data and each event causes big change. 
Also when data gets very large it will lose dynamic effect, yet balance changes can happen in server, maps, players etc. Downscale will retain dynamic change in bias. As SR calculation uses also map bias then i think it is good if map bias can work this way and refresh itself. 

**8. Change in bias downscale vs no downscale**

![closeup](https://user-images.githubusercontent.com/99591943/180621457-13e685ca-d361-4536-b5b0-0954c19c16d2.png)

Here is a 25% win graph for one team, zoom in of events 2500-2600. Orange line is downscale to 100

Downscale: Red marked circles show first -0,7% lose then about 10 wins with +0,2% and then 2 loses -0,7% so fluctuating around 2%.
it can go up by amounts of 0,2-0,3% and go down about 0,7% -0,8%
depending on real events so fluctuations can happen around 2-3% either way.  

No dowscale: old data that is stuck under a 0,1 line so it means over 100 games there is no change in bias. 

It is mathematically very easily achievable and can apply to old big numbers as well and down or upscale to desired value

**Reply by rmarquis** [reply]
Hi @u1ps, sorry for the long delayed answer.

I'm the main developer that worked on the rating system as implemented, though I'm barely involved with the game development anymore. I'm definitely seeing this as an excellent suggestion!

Extending the current algorithm with a "certainty decay" parameter  (for players but that's also valid for maps) has been [mentioned in the docs](https://etlegacy.readthedocs.io/en/latest/skillrating.html#advanced-questions). Do note the implementation of map parameters is done with a mean and certainty values, so I am not sure the above would be the best way to improve the current system (recording all but restricting the number of input values to the last 100 entries, as opposed to working with Baysian values like it's done for players), but doing so would certainly be a good start point.

**Reply by u1ps** [reply]
So the blue line is not accurate presentation of what is happening right now because i did not have the exact equation what it it is following at the time and still do not know it. First game seems to give +2,4% in bias and map that has been on server for long time seems to remain in very static number. Orange lines are accurate and graphs prove they can adept and reflect changes, make it more dynamic. If someone can provide the current equation for map bias can put that in graph for comparison.

**Reply by rmarquis** [reply]
The whole equation that takes map biais into account for player rating calculation is quite complex, but for the sake of simplicity I'll just focus on the map bias approximation (it's only a tiny part of the way map bias affect ratings).

The biggest inaccuracy in your modeling of the blue line above is that the starting fluctuation is currently taken care of by continuity correction. In short, before a map is played first, 20 fake map data are taken into account (10 victories for each side) and that data is gradually replaced by real data. As such, after the map is played for the first time, the algorithm consider +-2.38% bias only instead of the crazy and impossible value of +-50%. After the map is played 20 times, only real data is left and then the counter is gradually increased.

Avoiding big fluctuation is important as to not mess completely with the player rating calculation, but indeed having a static number is not good either. There's a need to find a balance, as to be as close as possible of the reality but avoiding to keep too old data that don't match the current map bias situation anymore.

The same situation happens for player: at the moment there is a "dynamic factor" injected in the uncertainty of player rating, as to help the rating value move up or down and avoid to be stuck with a static value. Currently that factor is static only (default in trueskill), so the drawback is that the system doesn't care if a player doesn't play for a long time, while in reality we should become more uncertain about the player rating value. Implementing a non linear "time decay" value by making that parameter non static is a way to improve the system here,

Similarly, we could implement a time decay for map on the same basis: instead of checking how many times a map is played, look at the last time it is played and adjust the bias from there. I am not sure how I'd implement this, but basing the system on time (as opposed to the number of times a map is played) seems to be a more robust implementation. After all, if we don't play a map for 6 months, the biais should be more incertain than a map that has been played just one hour ago. We don't actually collect the timestamp of the last time a map is played (we do for last players rating update), so that's one big change we might need to add in the database.

I'll try to think more about it, but any suggestion is welcome.

**Reply by u1ps** [reply]
> In short, before a map is played first, 20 fake map data are taken into account (10 victories for each side)

That is what i am suggesting endlessly use "fake map data". For example starting with 100 games would be 50 victories for each side. Next game will be 51/50 and this is where downscale needs to happen, change in ratio but sum will always remain 100. 

This will basically always display win ratio of last 100 matches only and keep refreshing the data.

> 
> Avoiding big fluctuation is important as to not mess completely with the player rating calculation, but indeed having a static number is not good either. 

Doubling the "100" value will decrease dynamic fluctuation 2x, so avoiding big fluctuation but enough to keep it dynamic.

100 is 0,500 change in map bias after each game
200 is 0,250
400 is 0,125
500 is 0,100

> 
> After all, if we don't play a map for 6 months, the biais should be more incertain than a map that has been played just one hour ago. 

switching values from 500 to 100 depending how often or when map was last played can decrease the change in bias from 0,1 to 0,5. (If this is wanted) Going as low as 20 will be 2,4% in change behaving like a new map.

But i think getting map bias more dynamic will be better already and it can be achieved very easily, without changing much, just few equations. Implementing any kind of time decay would be next step probably more complex.

**Reply by rmarquis** [reply]
> But i think getting map bias more dynamic will be better already and it can be achieved very easily, without changing much, just few equations. Implementing any kind of time decay would be next step probably more complex.

Yes, let's do that as a first step.

For the initial continuity correction, the value of 20 has been chosen because it is is the minimal value required for normal approximation of the Bernouilli distribution: N(np, np(1-p)) with n > 20 with p not near 0 or 1. That's basically what we want, to approximate the real map bias as fast as possible to get more accurate skill rating, while getting rid of most of the damaging fluctuation.

As such, I'd suggest to keep that initial continuity correction the way it is. Then, we can let grow map win/loss numbers to an appropriate value, after which we can apply "downscaling" and keep only the last N played values to compute the map bias.
To chose the value of N, we need to decide approximately how much time of data we need to keep.

Let's say a server is running a pool of maps 24/7. The map pool is maybe 15 different maps in average, and the duration of a map play could be 15 to 20 minutes in average (most maps play give 20 minutes of time, but they might get won by the attacking side earlier).

So let's say 15 minutes of play in average: that's 96 maps played a day, and 2922 maps a month (1 month being 365.25/12 days), so about ~3000 maps played during an entire month. With a map pool of 15 maps, that's 200 play per map in a month. As such, N = 200 might be a good first approximation to keep data of the last 30 days, and N = 100 good if we want to keep data of the past 2 weeks.

Does my quick math make sense?

In a nutshell, for 1 month of data:
* initial continuity correction of map = 20
* once all fake data is gone, let it grow to 200 plays
* downscale to 200 and ignore older data.

**Reply by u1ps** [reply]
>     * initial continuity correction of map = 20
> 
>     * once all fake data is gone, let it grow to 200 plays
> 
>     * downscale to 200 and ignore older data.

I was also thinking like that. This way everything is working exactly right now. 
New addition would be that once it reaches N value of games. It will start downscale and from that point on it will always act like its is N + 1th match. It will prevent the number getting infinitely large and it capable of refreshing data towards real balance state of map.

 So only need to write code to check number of games and when the value is over N, apply downscale. Also changing the number N in future can be used to increase or decrease the dynamic aspect if needed/wanted.
 
 If i find the excel where i simulated this i can post some pics of the graph for comparison with downscale applied starting from N and fixed initial continuity correction of map = 20
 
 And yeah the math seems about correct :D 200 - should be around month

**Comment by u1ps** [comment]
So this is more correct now. Blue line is close to how it is right now. Orange line will take over after 200th event.

It is 0,250% change in map bias in a 0,000 balance situation. This also means 4 wins in row will make +1,000% change in balance. 

When map balance is +25,000% for one team. it is 0,125% when win -0,375% when lose. Lose 4 games in row and map bias will be +23,500%

If this feels too much change, value 500 can be used. (0,250 becomes 0,100)

Also i think it would be nice to use 0,00 decimal place for map bias too, just like SR does. So it will reflect changes more accurately

![map bias](https://user-images.githubusercontent.com/99591943/218311167-ffc8927d-275a-4dab-896d-91ef6690e0d4.png)

**Reply by rmarquis** [reply]
1% for 4 consecutive wins sounds okay to me.

Do you mind sharing the equation for downscaling that you are using? I have one, but it requires some changes in the local database, and I wonder if there is something more straightforward so we could implement it quickly.

**Reply by u1ps** [reply]
deleted misleading equation

**Comment by u1ps** [comment]
![image](https://user-images.githubusercontent.com/99591943/218313020-49e5163a-a5ab-4bee-b06d-96e36d864dcd.png)
200 is the N value, C3 is 1 or 0 (win or lose), A50 is amount of wins in database for the team

**Comment by u1ps** [comment]
Here is the line in work:
![work](https://user-images.githubusercontent.com/99591943/218313426-64597032-6c3e-4db5-aea1-db0463502ae9.png)

at 200th game there should be for ecample120wins for one team 80 for other. This is the first previous value in last amount of total wins i was refering to

**Reply by u1ps** [reply]
btw what does it store exactly in local database? basically it has to increase value of wins in one team and lower the value in other team in right portion according to ratio to keep sum of wins always the same. As long as u can make change in those numbers with equation you are good to go

here is list of 20 wins for allies starting from 100wins for allies, 100 wins for axis. so after 20 games allies should have about 109,5 wins and axis about 90,50 wins. U can use this for reference to test if it is working correctly
![tgest](https://user-images.githubusercontent.com/99591943/218314278-4644b513-8e5a-4a62-b465-576b3e2925da.png)

**Reply by rmarquis** [reply]
Thank you.

So the issue is that the number of wins for each team is currently stored as integer in the SQLite database, while we now need float. We need to update the database schema and ensure the update is easy and seamless for admins.

Since the database schema needs to be updated, I'll try to think of what extra information we also need for the future enhancement with time decay (since we'd otherwise need to modify the schema again), i.e. adding timestamp of last time the map was played. Also I see timestamp of player rating are stored as text (instead of datetime) in the database, which is okay for debugging purpose but annoying when we need to reuse that info.

**Reply by u1ps** [reply]
If u get timestamp for last time the map was played. You could implement current ratio downscale to 20 after period of time. For example when map has 120 wins, 80 loses. Not played for a year for example. You can make those numbers 12 and 8. And this will bring bigger fluctuation, behaving as it is 21th game and and start climbing for that 200 mark again. 

Map vote system has last played value on map, dunno if that data can be used?

**Reply by rmarquis** [reply]
> If u get timestamp for last time the map was played. You could implement current ratio downscale to 20 after period of time.

This is an excellent suggestion! The ratio downscale could even be totally dynamic, scaling from 200 to 20 based on the elapsed time. Glicko algorithm has something similar for time decay, so we can just adapt what's existing and already proven here. So we just need to add timestamp to the database, nothing more.

Edit: Adding performance stats (to check what the accuracy of the team winning prediction is) might be good too.

> Map vote system has last played value on map, dunno if that data can be used?

Correct. The map vote system use a separate, file based storage system (instead of database). For the sake of simplicity I would keep both system separate at the moment, as:

1/ we need to use database for skill rating anyway, so better have all the required information in one location
2/ some people might want to disable the database/skill rating system at compilation time but might still want to use the map vote system
3/ the map vote system code could be rewritten later on to use the database (if desired).

**Comment by u1ps** [comment]
Here is simulation of 5000 games with correct graphs. Starting from balanced state. First 2500 games it is 25%win rate for 1 team. After 2500 25% win rate for other team. Orange line is N=200. So the blue line would behave like orange line after 200th event capable of adjusting to real time balance changed
![sim](https://user-images.githubusercontent.com/99591943/218323086-c785de5b-02d0-4c6a-b7e4-619319efc7e4.png)

**Comment by u1ps** [comment]
N200 vs N500 going from +5% win to -5% win at 2500th event
![n200](https://user-images.githubusercontent.com/99591943/218324335-46349bbf-def6-44f3-aa7d-42af56e9a40f.png)
![n500](https://user-images.githubusercontent.com/99591943/218324338-0abb5f86-4f0c-423a-acb9-13a9fddda5d7.png)

**Reply by rmarquis** [reply]
Looking good! I still feel N = 200 is more appropriate than N = 500, as 2500 datapoint is basically a full year of data. A drastic change in map win ratio like the above would still take about ~2 months to correct by itself, which seems adequate (with N = 500 we're looking at about ~5 to 6 months).

The only change I'd do is storing the map win ratio in the database normalized to 100 (and then multiplied by N = 200 in the code), so in case we want to change it later in the code, the stored data can be reused without any issue.

Also adding all useful data for the map vote system in the database, in case we fully switch to SQLite in the future (like number of time the map was played, number of other maps played since it was last played, so we can still derivate any info we might need later on such as the last period for the map to get a more precise and variable N).

**Reply by u1ps** [reply]
yeah, just providing some simulated data to see differences between them.

Both have like cons and pros :P

N200 = correction time 200 maps(about month); More fluctuation so possibility leaning towards opposite direction higher, but can also reflect change much faster.
N500 = correction time 500 maps(more than 2 months); about x2 less fluctuation so possibility leaning towards opposite direction smaller but will take longer time to adjust to change.

I also lean towards 200 as it can reflect change faster. Real life data will be slightly different as there will be more factors in play but simulating this is probably closest to real thing.

**Comment by u1ps** [comment]
Real live balance state is equal to AXIS[win%] + map bias % if i understood this correctly? so AXIS: 69.3+9.9=79.2%, ALLIES: 30.7-9.9=20,8% or those numbers are in different scale and can not be compared?

for example if allies have map bias +20% and real time win rate estimate is 30%. So does that mean this is balanced state for current moment? 20+30=50%

![balanceeee](https://user-images.githubusercontent.com/99591943/218330885-7f410f72-e28a-41a9-a229-f8a7c0ce1c34.png)

**Reply by rmarquis** [reply]
No, map bias is a component of live winning percentage.

If the win/loss ratio for a given map is 0.6 for axis (and thus 0.4 for allies), then the map bias is (0.6-0.4)/2 = 0.1 for axis, (0.4-0.6)/2 = -0.1 for allies.

Live winning percentage is a complex function that depends on map bias and rating of all players that are currently playing (or played the map but went to spec or left the server).

**Reply by u1ps** [reply]
So live winning percentage already takes into account map bias. What about color grading this value from 0 to 100 going from red to green. So balanced state is both 50.0 neutral (yellow) shifting towards 100 goes green, towards 0 red. So i assume this is the first number u want to see then to estimate balance so it should be more eye catching.

Map bias white maybe only with green/red +/- mark to indicate informative state of map based on history. just quick fill in paint to show what i mean(not perfect). This follows the principle of highlighting the most important number. I think color grading info this way might be slightly better?
![balanceeeenew](https://user-images.githubusercontent.com/99591943/218333760-7620c9df-cb1b-485f-bdd5-72b47576032f.png)

**Reply by rmarquis** [reply]
As a general rule I am against gradient colouring for the sake of it, as to avoid any christmas tree effect. In fact I would be more favourable to removing colour where possible, and keeping a few, consistent basic colours across the UI where important (such as timing).

When I implemented the rating code in UI, I did in fact try various colour alternative, but eventually opted for the less distracting way.  I've really only added colour to map bias values because they were too hard to distinguish from the +/- values, but the winning percentage is imho readable enough, not to mention the prediction itself is only that: a prediction, not a fact given by an oracle that is 100% accurate.

**Reply by u1ps** [reply]
So it is intentionally kept this way to cause less distraction as it is prediction. Opposite is to really just color the + and - mark on map bias to make it stand out less too, unless it becomes undistinguishable this way.

**Reply by rmarquis** [reply]
We can try that again, yes.

**Reply by rmarquis** [reply]
Just a quick heads-up. I still would like to improve the current system, but I have to conceed the amount of tests required to avoid the transition being painful for existing servers (and avoid a full reset of stats) kinda discourages me. Working on stats is fun, but getting angry players throwing rocks at you is much less fun.

Let me know if you're interested in handling this issue. We could design new fields needed for the database (for present and future features related to stats), before seeing what's needed to update schema painlessly. And after that only work on the present biais issue.

**Reply by u1ps** [reply]
not sure what else is wanted for future but to make this current suggestion work. Need number of wins for allies and axis for each map, that will change after game

**Reply by rmarquis** [reply]
From the discussion above, what would be needed is:

* adding win/loss numbers but as float, not integer
* storing timestamp of last time the map was played
* changing last time a player has its rating updated from text to timestamp (text might still be useful for quick debug purpose, so maybe keep it? Also add map last play time as text for the same reason)

I'd also check what stats the map voting system use, and add the required field to the database (so the system can be moved entirely to database later on).

**Reply by u1ps** [reply]
i have close to 0 experience with coding so someone with more experience is best fit to check over those 3 points stated by you.

**Reply by rmarquis** [reply]
Thank you. In the meantime, I've added a new ticket (#2317) with a summary of needed changes.

**Reply by u1ps** [reply]
One more notice not related to map bias exactly but SR gained after game. 

Game server has a peak time. This is how biggest % portion of players experience it. There are cases where you play with bots or few players and manage to get SR much easier compared to peak time of server. To fix this also add amount of players into account when calculating SR gain. 

For example if server has 40 slots max SR gain is 100%. When server has 20 people online it is 50%. 10 people online 25%. 

Things get more difficult with more players so it seems reasonable to reward players most in server peak times.

+0.20 sr gain in 10vs10 becomes +0.10, in 5vs5 +0.05.

Add multiplier based on player amount. from 1x to 0x. Or from 1x to 0.5x if want to lower SR gain with low amount  of players up to -50%. Depends how much you want to lower SR gain in matches with few people

**Reply by rmarquis** [reply]
No, SR gain or loss is scaled relatively to the number of players (in fact it is scaled to the relative team performance). There's indeed some simplification done in the sense that a map with 36 players or 6 players might play quite differently and be easier/more difficult to win, but that's not inherently related to the rating algorithm. In any case, the improvement you are suggesting above simply doesn't make any sense.

Also, please open a new discussion or a new ticket in the tracker if you want to discuss new improvements to SR. The suggestion for dynamic map bias are also tracked in #2317 (I don't particularly follow this repository nowadays).
